aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -236,6 +236,38 @@ SparseTensor& resize_as_sparse_(SparseTensor& self, const SparseTensor& src) {
   return self;
 }
 
+SparseTensor dense_to_sparse(const Tensor& self){
+  return dense_to_sparse(self, self.dim());
+}
+
+SparseTensor dense_to_sparse(const Tensor& self, int64_t sparseDims){
+  int64_t dims = self.dim();
+  AT_CHECK(sparseDims > 0, "sparseDims must be >0");
+  AT_CHECK(sparseDims <= dims, 
+    "sparseDims must be less than or equal to self.dim()");
+  at::TensorOptions sparseOpts = self.type().toSparse().options();
+  std::vector<int64_t> sizes = self.sizes().vec();
+
+  LongTensor indices = self.nonzero().transpose(0, 1);
+  std::vector<int64_t> maskShape;
+  if(sparseDims != dims){
```

Comment:
for case of sparseDims == dims, will it be better to use advanced indexing?
```
Dense = torch.randn(3, 4, 5, 6)
indices = Dense.nonzero().t()
values = Dense[indices.chunk(indices.size(0))].squeeze_()
return torch.sparse_coo_tensor(indices, values, Dense.shape)
```

Reply:
similarly, for sparseDims != dims, you can
```
I = D.nonzero()[:, :shapeDims]
uniq_I = I.unique(dim=0).t()
values = D[uniq_I.chunk(sparseDims)].squeeze_()
return torch.sparse_coo_tensor(uniq_I, values, D.shape)
```

Reply:
@weiyangfb ok, lmk if this looks good.  I went with ```narrow``` instead of ```index``` in one place because the API in C++ is a little cleaner and re: convo in general slack its no less efficient.